### DISCUSSION TITLE
다수의 혹은 모든 Source에 대한 데이터 Publish, 검색을 처리하는 방법에 대한 검토 요청

### DISCUSSION THREAD

**Discussion by syncpark** [aicers/giganto#1283] [Ideas]
Giganto에서 Search & Publish 할때 Source를 전체로 지정하는 요청에 대한 이슈가 있습니다.
이에 대한 해결 방법이 필요해서 고민해봤으니 검토 부탁합니다.

### REconverge -> Kafka 데이터 요청의 처리
* REconverge는 Kafka의 Topic을 지정해서 데이터를 요청한다.
* Kafka는 REconverge의 요청에 따라 데이터를 전송하고 Consumer(REconverge)와 마지막으로 읽은 위치(Offset)을 기록해둔다.
* 다시 REconverge가 데이터를 요청하면 Kafka는 Consumer를 확인하고, 마지막 위치 다음부터 데이터를 전송한다.

### REconverge -> Giganto 데이터 요청의 처리
* REconverge는 아래 Message 구조체의 내용(수집 장비 이름, 데이터 분류 코드, 시작시간, 종료시간, 건수)을 지정해서 데이터를 요청한다.
* REconverge는 Giganto로부터 마지막으로 수신한 데이터를 기준으로로 Message 구조체의 내용을 수정해서 요청한다.
```Rust
struct Message {
    source: String,
    kind: String,
    start: i64,
    end: i64,
    count: usize,
}
```

### 문제점
* REconverge가 Giganto로부터 읽은 마지막 위치를 기억하고 있어야 한다.
  이 값을 어딘가에 보관해두지 않으면 같은 데이터를 반복해서 분석하게 된다.
  Giganto보다는 REconverge가 재시작할 가능성이 매우 높기 때문에 REconverge가 이 값을 보관하는 것은 피하는 것이 좋다.
* Message 구조체의 `end` 값은 REconverge에게는 의미가 없지만, 현재 Giganto에 데이터 요청할 때는 필수적으로 지정해야 한다. 
  REconverge는 지속적인 분석이 기본 동작이므로 `end` 필드값이 사용될 일이 없다.
* REconverge가 `source` 필드 값으로 모든 수집장비(All)를 지정했을 때 이를 처리하기 위해서는 Giganto의 검색 방식에 수정이 필요하다.

### 해결 방법: Offset 단위로 이벤트를 관리
* 수집한 데이터들을 `source`에 관련없이 들어오는 순서대로 1000 혹은 10000건 단위(하나의 Offset)로 키값(Source+시간)들을 묶어서 저장한다.
* Offset의 Key는 포함된 첫번째 이벤트의 시간 값으로한다.
* REconverge에서 요청하는 데이터는 Offset 단위로 전송한다.
* 전송 대상인 REconverge의 id(agent-id@host-id)와 offset 값을 기억하고 있다가, 이후 전송 요청이 오면 기억한 위치에서 시작한다.
* Source를 지정한 검색에는 기존 검색 방식을 적용, 모든 Source를 지정하면 Offset 단위로 저장한 방식을 적용한다.
* 다수의 REconverge가 같은 종류의 데이터를 분산 처리한다면 대상의 ID(예: Kafka의 Consumer)값만 통일하면 된다.

**Comment by msk** [comment]
> * REconverge가 Giganto로부터 읽은 마지막 위치를 기억하고 있어야 한다.
>   이 값을 어딘가에 보관해두지 않으면 같은 데이터를 반복해서 분석하게 된다.
>   Giganto보다는 REconverge가 재시작할 가능성이 매우 높기 때문에 REconverge가 이 값을 보관하는 것은 피하는 것이 좋다.

이 결론은 잘 이해가 안 가는데... REconverge가 그 값을 모델에 넣어두면 되지 않나요?

**Comment by syncpark** [comment]
모델은 배포 가능합니다. 지금 진행중인 프로젝트 2개에도 모두 그런 요건이 들어있습니다.
모델에 해당 정보를 기록하면, 모델을 여러 분석장비에 배포하여 분석하는 경우 모두 같은 데이터를 분석하게 됩니다.

**Comment by octoaide** [comment]
Hello @sophie-cluml and @syncpark,

This issue, `#181`, proposes a method for handling data publishing and searching for multiple/all sources, specifically suggesting an 'Offset unit' approach. It's been open for a while and the discussion around the 'last read position' was never fully resolved.

Could you please take a look at the proposed solution in the issue description and let us know if this approach is still considered viable or if there have been any new developments/priorities that impact this? We'd appreciate your thoughts on the best way to move this forward. Thanks!

**Comment by sophie-cluml** [comment]
This task appears to have open points for discussion, so I will convert this to a discussion.
